### ISSUE TITLE
utils.reset(net) resets hidden states in unrelated networks

### ISSUE THREAD

**Issue by GraceDyer** [jeshraghian/snntorch#438]
**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**snnTorch version**

`1.0.0`

Also reproduced on the current `master` commit [`5fde1b8e`](https://github.com/jeshraghian/snntorch/commit/5fde1b8e6039792484eb32c3ac57174516dd9cf4).

**Description**

I found a reproducible bug in the public helper `snntorch.utils.reset(net)`.

The helper appears intended to reset hidden states belonging to the supplied network, but calling `utils.reset(net_a)` also clears hidden states in a completely separate `net_b` when both networks contain the same neuron class.

The two networks in the reproduction below do not share modules, tensors, or parameters.

From source inspection, `utils.reset(net)` first uses the supplied network to determine which neuron classes it contains. It then calls class-level helpers such as `snn.Leaky.reset_hidden()`. Those helpers iterate over the shared neuron-instance registry, so they reset every live `Leaky` instance rather than only the instances reachable from the supplied network.

Because of this, resetting one model can silently modify the temporal state of another independent model in the same process.

I reproduced this with both snnTorch `1.0.0` and the current master commit in fresh CPU-only Docker containers using Debian GNU/Linux 12, Python 3.11.16, and PyTorch 2.8.0+cpu.

This is distinct from [PR #414](https://github.com/jeshraghian/snntorch/pull/414): that change fixed recursive discovery of nested neurons, while the reset operation itself remains global rather than local to the supplied network.

**Minimal code to reproduce the error/bug**

```python
import torch
import torch.nn as nn
import snntorch as snn
from snntorch import utils

net_a = nn.Sequential(
    snn.Leaky(beta=0.5, threshold=10.0, init_hidden=True)
)
net_b = nn.Sequential(
    snn.Leaky(beta=0.5, threshold=10.0, init_hidden=True)
)

# Give each independent network a different nonzero membrane state.
_ = net_a(torch.tensor([0.6]))
_ = net_b(torch.tensor([0.7]))

net_a_before = net_a[0].mem.clone()
net_b_before = net_b[0].mem.clone()

# Only net_a is passed to reset().
utils.reset(net_a)

net_a_after = net_a[0].mem.clone()
net_b_after = net_b[0].mem.clone()

print("net_a_before =", net_a_before.item())
print("net_a_after =", net_a_after.item())
print("net_b_before =", net_b_before.item())
print("net_b_after =", net_b_after.item())
print("net_b_unchanged =", bool(torch.equal(net_b_after, net_b_before)))
```

**What is expected**

Calling `utils.reset(net_a)` should reset only the hidden states belonging to `net_a`.

The membrane state in the unrelated `net_b` should remain unchanged.

**What actually happens**

Both networks are reset, even though only `net_a` is passed to `utils.reset`.

Example output from the Docker reproduction:

```text
net_a_before = 0.6000000238418579
net_a_after = 0.0
net_b_before = 0.699999988079071
net_b_after = 0.0
net_b_unchanged = False
```

This looks like an implementation bug rather than an environment issue because the reset helper dispatches to class-level methods that operate on the shared global instance registry.

**Comment by ethanmiller-dev** [comment]
Thanks for posting this! I’ve been trying to figure out why my model’s accuracy wouldn’t improve, and this finally explains it.

**Comment by ixfd64** [comment]
I believe #449 should fix this. Let us know if there are still issues.
